Pivot table support in the backend

robdaemon opened the pull request:
Working on generic pivot table support for (all?) databases. This generates the relevant                                         
breakouts for the pivot table queries, then runs the queries serially, returning the result.

Commit:
pivot tables

add new API

add driver flag for pushdown of pivot tables

Commit:
pivot table support in the backend

Working on generic pivot table support for (all?) databases. This generates the relevant
breakouts for the pivot table queries, then runs the queries serially, returning the result.

[ci all]

robdaemon commented on src/metabase/driver.clj:
```diff
@@ -412,7 +412,10 @@
+    :percentile-aggregations
+
+    ;; Does this driver support pivot table pushdown? (executing a pivot table in a driver-specific way)
+    :pivot-table-pushdown})
```
this is a placeholder for now that we can use later

sbelak replied:
A better name would be some variation on `OLAP` as we'll eventually want to expose this beyond just pivot tables.

robdaemon commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -0,0 +1,29 @@
+  (api/read-check Database database)
+
+  ;;TODO: make this use process-query-async, and run all four queries in parallel
+  (let [pivot-queries (pivot/generate-queries query)]
```
the calls to `pivot/` likely belong somewhere else. Do we want to have a `:type :pivot` like we have `:type :query` now and such? that would move the work of rewriting a query out of this API endpoint and into the query processor.

Commit:
fix unit test for pivot table

Writes as proper of a unit test as possible for this - the number of rows returned in this pivot table request are entirely too large to do a full equality assertion.

sbelak commented on src/metabase/api/advanced_computation.clj:
```diff
@@ -0,0 +1,29 @@
+         :as        query} :body}]
+  {database (s/maybe s/Int)}
+
+  (when-not database
```
Can this be done with Schema? Seems kind of silly to use s/maybe and then throw

sbelak commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -0,0 +1,20 @@
+(defn- generate-breakouts
+  "Generate the combinatorial breakouts for a given query pivot table query"
+  [breakouts]
+  (when (not= 3 (count breakouts))
```
I don't see why we have to enforce 3 breakouts. Having a total seems useful for 1-oo breakouts

sbelak commented on src/metabase/query_processor/pivot.clj:
```diff
@@ -0,0 +1,20 @@
+  (when (not= 3 (count breakouts))
+    (throw (ex-info (trs "A pivot table query requires three breakouts") {:breakout breakouts})))
+
+  [(take 1 breakouts) (take 2 breakouts) (take-last 2 breakouts) []])
```
This would be nicer/more generically expressed as a powerset problem (ie. generate all subsets).

Commit:
starting to add UNION to mbql

based on talking to sbelak and paulrosenzweig today, adding a UNION type to MBQL is the best approach to handling this for all SQL-derived datasources

robdaemon commented:
replaced by metabase/metabase#13946
